### PR TITLE
Netresult memory size optimization

### PR DESCRIPTION
**Pull request by ihavnoid**
Instead of keeping a list of std::pair<float,int>, just keep a flat array of floats.  This reduces per-entry NNCache size by half, hence allowing bigger cache sizes for larger playouts.

### CODE REVIEW

File: src/Network.cpp [hunk 1]
```diff
@@ -930,45 +930,33 @@ Network::Netresult Network::get_scored_moves_internal(
...
-                [&vtx](scored_node const& test_item) {
-                return test_item.second == vtx;
-            });
-
             auto score = 0;
-            // Non-empty squares won't be scored
-            if (item != moves.cend()) {
-                score = int(item->first * 1000);
-                assert(vtx == item->second);
+            auto vertex = state->board.get_vertex(x, y);
```
Comment (Hersmunch) [inline src/Network.cpp]: const was removed

File: src/Network.cpp [hunk 2]
```diff
@@ -981,12 +969,22 @@ void Network::show_heatmap(const FastState* const state,
...
-    assert(result.first.back().second == FastBoard::PASS);
-    const auto pass_score = int(result.first.back().first * 1000);
+    const auto pass_score = int(result[BOARD_SQUARES] * 1000);
     myprintf("pass: %d\n", pass_score);
-    myprintf("winrate: %f\n", result.second);
+    myprintf("winrate: %f\n", result[BOARD_SQUARES+1]);
 
     if (topmoves) {
+        std::vector<std::pair<float,int>> moves;
+        for (int i=0; i < BOARD_SQUARES; i++) {
```
Comment (Hersmunch) [inline src/Network.cpp]: auto

File: src/Network.cpp [hunk 3]
```diff
@@ -981,12 +969,22 @@ void Network::show_heatmap(const FastState* const state,
...
+            auto x = i % BOARD_SIZE;
+            auto y = i / BOARD_SIZE;
+            auto vertex = state->board.get_vertex(x, y);
```
Comment (Hersmunch) [inline src/Network.cpp]: Can all be const

File: src/UCTNode.cpp [hunk 4]
```diff
@@ -81,22 +81,31 @@ bool UCTNode::create_children(std::atomic<int>& nodecount,
...
-        if (state.is_move_legal(to_move, vertex)) {
-            nodelist.emplace_back(node);
-            legal_sum += node.first;
+    for (auto i = 0; i < BOARD_SQUARES+1; i++) {
+        if (i < BOARD_SQUARES) {
+            const auto x = i % BOARD_SIZE;
+            const auto y = i / BOARD_SIZE;
+            const auto vertex = state.board.get_vertex(x, y); 
+            if (state.is_move_legal(to_move, vertex)) {
+                auto p = std::make_pair(raw_netlist[i], vertex);
```
Comment (Hersmunch) [inline src/UCTNode.cpp]: Can be const. Same for similar line a few lines down

**Comment by ihavnoid** [comment]
Addressed code review comments.

File: src/UCTNode.h [hunk 5]
```diff
@@ -88,7 +88,7 @@ class UCTNode {
         ACTIVE
     };
     void link_nodelist(std::atomic<int>& nodecount,
-                       std::vector<Network::scored_node>& nodelist,
+                       std::vector<std::pair<float,int>>& nodelist,
```
Comment (gcp) [inline src/UCTNode.h]: Why remove the typedef?

File: src/Network.cpp [hunk 6]
```diff
@@ -930,45 +930,33 @@ Network::Netresult Network::get_scored_moves_internal(
...
-            if (state->board.get_square(rot_vtx) == FastBoard::EMPTY) {
-                result.emplace_back(outputs[idx], rot_vtx);
-            }
+            result[rot_idx] = outputs[idx];
         } else {
-            result.emplace_back(outputs[idx], FastBoard::PASS);
+            result[idx] = outputs[idx];
         }
     }
+    result[BOARD_SQUARES+1] = winrate_sig;
```
Comment (gcp) [inline src/Network.cpp]: This is very ugly. At least make it a pair of an array and a float, or define a type/class/struct for the result.

File: src/Network.h [hunk 7]
```diff
@@ -39,8 +39,9 @@ class Network {
     };
     using BoardPlane = std::bitset<BOARD_SQUARES>;
     using NNPlanes = std::vector<BoardPlane>;
-    using scored_node = std::pair<float, int>;
-    using Netresult = std::pair<std::vector<scored_node>, float>;
+
+    // {19x19 board positions, pass, winrate}
+    using Netresult = std::array<float, BOARD_SQUARES+2>;
```
Comment (gcp) [inline src/Network.h]: Rather use a custom struct/class or at worst a tuple.

**Comment by ihavnoid** [comment]
Addressed @gcp 's comments, although I am not sure it is satisfactory.  The thing is, because the loop index itself is the location of the move, I ended up using the index itself anyway and hence the 'foreach' style of loop iteration wasn't that useful anyway.

One additional change I made was that the policy is not raw std::array, it's now a std::vector sized of BOARD_SQUARE - this is because there are many locations doing std::move, and using a std::vector instead of std::array will avoid many deep copy operations.

File: src/Network.cpp [hunk 8]
```diff
@@ -981,12 +967,23 @@ void Network::show_heatmap(const FastState* const state,
...
     }
-    assert(result.first.back().second == FastBoard::PASS);
-    const auto pass_score = int(result.first.back().first * 1000);
+    const auto pass_score = int(result.policy_pass * 1000);
     myprintf("pass: %d\n", pass_score);
-    myprintf("winrate: %f\n", result.second);
+    myprintf("winrate: %f\n", result.winrate);
 
     if (topmoves) {
+        using ScoreVertexPair = std::pair<float,int>;
```
Comment (gcp) [inline src/Network.cpp]: You don't need to redefine this?

File: src/UCTNode.cpp [hunk 9]
```diff
@@ -81,24 +81,32 @@ bool UCTNode::create_children(std::atomic<int>& nodecount,
...
-        auto vertex = node.second;
+    for (auto i = 0; i < BOARD_SQUARES; i++) {
+        const auto x = i % BOARD_SIZE;
+        const auto y = i / BOARD_SIZE;
+        const auto vertex = state.board.get_vertex(x, y); 
         if (state.is_move_legal(to_move, vertex)) {
-            nodelist.emplace_back(node);
-            legal_sum += node.first;
+            const auto p = std::make_pair(raw_netlist.policy[i], vertex);
+            nodelist.emplace_back(p);
```
Comment (gcp) [inline src/UCTNode.cpp]: I think you can just do emplace_back(raw_netlist.policy[i], vertex)?

File: src/UCTNode.cpp [hunk 10]
```diff
@@ -81,24 +81,32 @@ bool UCTNode::create_children(std::atomic<int>& nodecount,
...
+            legal_sum += raw_netlist.policy[i];
         }
     }
+    {
+        const auto p = std::make_pair(raw_netlist.policy_pass, FastBoard::PASS);
+        nodelist.emplace_back(p);
```
Comment (gcp) [inline src/UCTNode.cpp]: Same.

File: src/Network.h [hunk 11]
```diff
@@ -39,8 +39,13 @@ class Network {
     };
     using BoardPlane = std::bitset<BOARD_SQUARES>;
     using NNPlanes = std::vector<BoardPlane>;
-    using scored_node = std::pair<float, int>;
-    using Netresult = std::pair<std::vector<scored_node>, float>;
+
+    // {19x19 board positions, pass, winrate}
+    struct Netresult {
+        std::vector<float> policy = std::vector<float>(BOARD_SQUARES);
```
Comment (gcp) [inline src/Network.h]: Gah, sucks how uniform initialization and initializer lists interact :(

**Comment by gcp** [comment]
>The thing is, because the loop index itself is the location of the move, I ended up using the index itself anyway and hence the 'foreach' style of loop iteration wasn't that useful anyway.

I understand, it was more about stuffing the extra info into magic locations. The struct is much cleaner.

File: src/Network.cpp (hunk 8 above)
Comment (gcp) [reply src/Network.cpp]: I'd reuse the typedef (even if the Vertex is not really the same), and put it wherever is most appropriate.

**Comment by ihavnoid** [comment]
Pushed a commit addressing the third batch of comments.

**Pull request merged**